File: tests/fixtures/project_java/src/main/java/auth/Jwt.java
package auth;

import java.util.Base64;
import java.util.Map;
import java.util.HashMap;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import users.User;

public class Jwt {
    private static final String SECRET = "stub-secret-key";
    private static final String ALGORITHM = "HmacSHA256";

    /** Signs a JWT token from the given payload map. Returns the encoded token string. */
    public static String signToken(Map<String, Object> payload) {
        if (payload == null) {
            throw new IllegalArgumentException("payload cannot be null");
        }
        if (payload.isEmpty()) {
            throw new IllegalArgumentException("payload cannot be empty");
        }

        String header = Base64.getEncoder().encodeToString("{\"alg\":\"HS256\",\"typ\":\"JWT\"}".getBytes());
        String body = Base64.getEncoder().encodeToString(payload.toString().getBytes());
        String signature = hmacSign(header + "." + body);
        return header + "." + body + "." + signature;
    }

    /** Decodes and verifies a JWT token. Returns the decoded payload. */
    public static Map<String, Object> decodeToken(String token) {
        if (token == null || token.isEmpty()) {
            throw new IllegalArgumentException("token cannot be null or empty");
        }

        String[] parts = token.split("\\.");
        if (parts.length != 3) {
            throw new IllegalArgumentException("invalid token format");
        }

        String expectedSig = hmacSign(parts[0] + "." + parts[1]);
        if (!expectedSig.equals(parts[2])) {
            throw new SecurityException("invalid token signature");
        }

        String decoded = new String(Base64.getDecoder().decode(parts[1]));
        Map<String, Object> result = new HashMap<>();
        result.put("raw", decoded);
        return result;
    }

    public static boolean isTokenExpired(Map<String, Object> payload) {
        if (payload == null) {
            return true;
        }
        Object exp = payload.get("exp");
        if (exp instanceof Long) {
            return (Long) exp < System.currentTimeMillis() / 1000;
        }
        return false;
    }

    private static String hmacSign(String data) {
        try {
            Mac mac = Mac.getInstance(ALGORITHM);
            mac.init(new SecretKeySpec(SECRET.getBytes(), ALGORITHM));
            byte[] hash = mac.doFinal(data.getBytes());
            return Base64.getUrlEncoder().withoutPadding().encodeToString(hash);
        } catch (Exception e) {
            throw new RuntimeException("HMAC signing failed", e);
        }
    }
}

File: tests/fixtures/project_go/auth/jwt.go
package auth

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

// TokenPayload represents the decoded contents of a JWT token.
type TokenPayload struct {
	UserID    int       `json:"user_id"`
	Email     string    `json:"email"`
	ExpiresAt time.Time `json:"expires_at"`
}

const jwtSecret = "stub-secret-key"

// SignToken creates a signed JWT token from the given payload map.
func SignToken(payload map[string]interface{}) (string, error) {
	if payload == nil {
		return "", fmt.Errorf("payload cannot be nil")
	}

	header := map[string]string{"alg": "HS256", "typ": "JWT"}
	headerJSON, err := json.Marshal(header)
	if err != nil {
		return "", fmt.Errorf("failed to marshal header: %w", err)
	}

	payload["iat"] = time.Now().Unix()
	if _, ok := payload["exp"]; !ok {
		payload["exp"] = time.Now().Add(24 * time.Hour).Unix()
	}

	payloadJSON, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("failed to marshal payload: %w", err)
	}

	headerB64 := base64.RawURLEncoding.EncodeToString(headerJSON)
	payloadB64 := base64.RawURLEncoding.EncodeToString(payloadJSON)
	signingInput := headerB64 + "." + payloadB64

	mac := hmac.New(sha256.New, []byte(jwtSecret))
	mac.Write([]byte(signingInput))
	signature := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))

	return signingInput + "." + signature, nil
}

// DecodeToken decodes and verifies a JWT token string, returning its payload.
func DecodeToken(token string) (*TokenPayload, error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, fmt.Errorf("invalid token format")
	}

	signingInput := parts[0] + "." + parts[1]
	mac := hmac.New(sha256.New, []byte(jwtSecret))
	mac.Write([]byte(signingInput))
	expectedSig := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))

	if !hmac.Equal([]byte(parts[2]), []byte(expectedSig)) {
		return nil, fmt.Errorf("invalid token signature")
	}

	payloadJSON, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("failed to decode payload: %w", err)
	}

	var raw map[string]interface{}
	if err := json.Unmarshal(payloadJSON, &raw); err != nil {
		return nil, fmt.Errorf("failed to unmarshal payload: %w", err)
	}

	result := &TokenPayload{}
	if uid, ok := raw["user_id"].(float64); ok {
		result.UserID = int(uid)
	}
	if email, ok := raw["email"].(string); ok {
		result.Email = email
	}
	if exp, ok := raw["exp"].(float64); ok {
		result.ExpiresAt = time.Unix(int64(exp), 0)
	}

	if result.ExpiresAt.Before(time.Now()) {
		return nil, fmt.Errorf("token expired")
	}

	return result, nil
}
